mintpy/createmask_cp.py

```diff
@@ -0,0 +1,142 @@
+#!/usr/bin/env python
+# Estimate fading signal in level-n time-series
+
+import os, sys
+import h5py
+from mintpy.objects import ifgramStack
+import numpy as np
+from mintpy.utils import readfile, writefile, ptime, utils as ut
+from mintpy import ifgram_inversion as ifginv
+import argparse
+import matplotlib.pyplot as plt
+
+def create_parser():
+    parser = argparse.ArgumentParser(description = 'Create an indication map for closure phase bias')
+    # input dataset
+    parser.add_argument('-i','--ifgramstack',type = str, dest = 'ifgram_stack',help = 'interferogram stack file that contains the unwrapped phases')
+    parser.add_argument('--nl', dest = 'nl', type = int, default = 20, help = 'connection level that we are correcting to (or consider as no bias)')
+    parser.add_argument('--maxMemory', dest = 'max_memory', type = float, default = 8, help = 'max memory to use in GB')
+    parser.add_argument('-o', dest = 'outdir', type = str, default = '.', help = 'output file directory')
+    return parser
+
+def cmd_line_parse(iargs=None):
+    parser = create_parser()
+    inps = parser.parse_args(args = iargs)
+    return inps
+
+# Obtain sum of consecutive sequential closure phase of connection n
+def cum_seq_closurePhase(inps,n,box): # we call triplets 2-level closure phase.
+    cp_idx = []
+    SLC_list = inps.SLC_list
+    for i in range(inps.NSLC-n):
+        ifgram = []
+        flag = True
+        for j in range(n):
+            ifgram.append('{}_{}'.format(inps.SLC_list[i+j],inps.SLC_list[i+j+1]))
+        ifgram.append('{}_{}'.format(SLC_list[i],SLC_list[i+n]))
+        for ifgram_name in ifgram:
+            if ifgram_name not in inps.date12_list_all:
+                # print('Interferogram not found: ', ifgram_name )
+                flag = False
+        if flag:
+            cp_idx.append([inps.date12_list_all.index(ifgram[j]) for j in range(n+1)])
+
+    cp_idx = np.array(cp_idx, np.int16)
+    cp_idx = np.unique(cp_idx, axis = 0)
+    
+    num_cp = len(cp_idx)
+    print('Number of closure measurements expected, ', len(SLC_list)-n)
+    print('Number of closure measurements found, ', num_cp)
+    
+    if num_cp <1:
+        print('No closure phase measurements found, abort')
+        sys.exit()
+        
+    box_width  = box[2] - box[0]
+    box_length = box[3] - box[1]
+    phase = readfile.read(inps.ifgram_stack, box=box,print_msg=False)[0]
+    cum_cp = np.zeros((box_length, box_width), np.complex64)
+    for i in range(num_cp):
+        cp0_w = np.zeros ((box_length, box_width), np.float32)
+        for j in range(n):
+                    idx = cp_idx[i,j]
+                    cp0_w = cp0_w + phase[idx,:,:] - inps.ref_phase[idx]
+        idx = cp_idx[i,n]
+        cp0_w = cp0_w - (phase[idx,:,:]-inps.ref_phase[idx])
+        cum_cp = cum_cp + (np.exp(1j*cp0_w))
```

Comment:
I would like to understand what is going on here. 
What is the exact quantity that we want to compute here? Isn't it the sum of the wrapped closure phases? 
It looks like you do compute the closure phases from unwrapped phases first and then you sum the complex values. I'm not sure if the angle of the sum of the complex values could be called the cumulative closure phase. 

If you really want the sum of the closure phases I assume you could keep adding the unwrapped closure phases and then at the end you could wrap it to get rid of unwrapping errors contribution to closure phase. 

If no that is not what you need would you please clarify and put in some comments describing this quantity (i.e., cum_cp).

Reply:
ok sounds good and consistent with the manuscript.